### ISSUE TITLE
Calling cloud function throws an internal error with no details - `FirebaseFunctionsException [firebase_functions/17999]- An internal error has occurred, print and inspect the error details for more information.`

### ISSUE THREAD

**Issue by hspinks** [FirebaseExtended/flutterfire#8037]
I'm trying to call a very simple cloud function from my Flutter app following the [FlutterFire docs](https://firebase.flutter.dev/docs/functions/usage#calling-endpoints), but I'm getting a pretty unreadable error. The error occurs whether I'm using the emulators, or attempting to call the function fully deployed with Firebase.

My cloud function code:

```typescript
import * as functions from "firebase-functions";

export const ping = functions.https.onCall(() => ({ ping: "pong" }));

```

And my Flutter code in `main.dart`:

```dart
import 'package:cloud_functions/cloud_functions.dart';
import 'package:firebase_core/firebase_core.dart';
import 'package:flutter/material.dart';
import 'package:my_app/firebase_options.dart';
import 'package:my_app/my_app.dart';

void main() async {
  WidgetsFlutterBinding.ensureInitialized();
  await Firebase.initializeApp(options: DefaultFirebaseOptions.currentPlatform); // Uses options configured by FlutterFire CLI

  // Does not work regardless of whether this line is commented out
  // FirebaseFunctions.instance.useFunctionsEmulator('localhost', 5001);

  HttpsCallable ping = FirebaseFunctions.instance.httpsCallable('ping');
  await ping(); // Exception thrown on this line

  runApp(const MyApp());
}
```

When running this code, I get a `FirebaseFunctionsException ([firebase_functions/17999] An internal error has occurred, print and inspect the error details for more information.` error:[![Exception details][1]][1]

Even as I dive deeper into the call stack, I'm not able to get any helpful information. Setting a breakpoint in the `call` method of [MethodChannelHttpsCallable](https://github.com/FirebaseExtended/flutterfire/blob/82c1fe111d00b499e9587d235e4499b03cfe17bd/packages/cloud_functions/cloud_functions_platform_interface/lib/src/method_channel/method_channel_https_callable.dart#L21) shows me that the exception is thrown when calling the `MethodChannelFirebaseFunctions.channel.invokeMethod` method, but I can't see any immediately obvious reason why that would be failing.

I'm using other FlutterFire libraries in my app (specifically Auth and Firestore), with no problems at all, both with the emulator and the deployed Firebase project.

Here's my `pubspec.yaml`:

```yaml
name: <redacted>
description: <redacted>
version: <redacted>
publish_to: none

environment:
  sdk: ">=2.15.1 <3.0.0"

dependencies:
  cloud_firestore: ^3.1.7
  cloud_functions: ^3.2.6
  firebase_auth: ^3.3.6
  firebase_core: ^1.12.0
  flutter:
    sdk: flutter
  flutterfire_ui: ^0.3.2

dev_dependencies:
  flutter_lints: ^1.0.0
  flutter_test:
    sdk: flutter

flutter:
  uses-material-design: true
```

Any help you can provide would be much appreciated!

  [1]: https://i.stack.imgur.com/d5Oy5.png

**Comment by darshankawar** [comment]
@hspinks 
Can you provide `flutter doctor -v` and the complete error log in text format ?
Also, can you try the plugin [example](https://github.com/FirebaseExtended/flutterfire/tree/master/packages/cloud_functions/cloud_functions/example) with your setup and see if you still get same error ?

**Comment by hspinks** [comment]
@darshankawar thanks for the quick response. I implemented the example you linked to as precisely as I could, with slight modifications to abide by my linter.

My Flutter client's MyApp is identical to the example you linked to, copy-pasted. My backend project's `functions/src/index.ts`:

```typescript
import * as functions from "firebase-functions";

export const listFruit = functions.https.onCall(() => {
  return ["Apple", "Banana", "Cherry", "Date", "Fig", "Grapes"];
});
```

The error output is:

```
flutter: [firebase_functions[/17999]()] An internal error has occurred, print and inspect the error details for more information.
#0      StandardMethodCodec.decodeEnvelope
package:flutter/…/services/message_codecs.dart:607
#1      MethodChannel._invokeMethod
package:flutter/…/services/platform_channel.dart:167
<asynchronous suspension>
#2      MethodChannelHttpsCallable.call
package:cloud_functions_platform_interface/…/method_channel/method_channel_https_callable.dart:23
<asynchronous suspension>
#3      HttpsCallable.call
package:cloud_functions/src/https_callable.dart:35
<asynchronous suspension>
#4      _MyAppState.build.<anonymous closure>.<anonymous closure>
package:my_app/app.dart:42
<asynchronous suspension>
```

The output of `flutter doctor -v` is:

```
[✓] Flutter (Channel stable, 2.8.1, on macOS 12.1 21C52 darwin-arm, locale en-US)
    • Flutter version 2.8.1 at /Users/<redacted>/Develop/sdks/flutter
    • Upstream repository https://github.com/flutter/flutter.git
    • Framework revision 77d935af4d (8 weeks ago), 2021-12-16 08:37:33 -0800
    • Engine revision 890a5fca2e
    • Dart version 2.15.1

[✓] Android toolchain - develop for Android devices (Android SDK version 31.0.0)
    • Android SDK at /Users/<redacted>/Library/Android/sdk
    • Platform android-31, build-tools 31.0.0
    • Java binary at: /Applications/Android Studio.app/Contents/jre/Contents/Home/bin/java
    • Java version OpenJDK Runtime Environment (build 11.0.10+0-b96-7249189)
    • All Android licenses accepted.

[✓] Xcode - develop for iOS and macOS (Xcode 13.2.1)
    • Xcode at /Applications/Xcode.app/Contents/Developer
    • CocoaPods version 1.11.2

[✓] Chrome - develop for the web
    • Chrome at /Applications/Google Chrome.app/Contents/MacOS/Google Chrome

[✓] Android Studio (version 2020.3)
    • Android Studio at /Applications/Android Studio.app/Contents
    • Flutter plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/9212-flutter
    • Dart plugin can be installed from:
      🔨 https://plugins.jetbrains.com/plugin/6351-dart
    • Java version OpenJDK Runtime Environment (build 11.0.10+0-b96-7249189)

[✓] VS Code (version 1.64.0)
    • VS Code at /Applications/Visual Studio Code.app/Contents
    • Flutter extension version 3.34.0

[✓] Connected device (2 available)
    • iPhone 13 (mobile) • <redacted> • ios            • com.apple.CoreSimulator.SimRuntime.iOS-15-2 (simulator)
    • Chrome (web)       • chrome                               • web-javascript • Google Chrome 98.0.4758.80

• No issues found!
```

**Comment by darshankawar** [comment]
Thanks for the update. I am wondering if you need to run it specifically on `FunctionsEmulator`. Can you uncomment the below line and try ?

> ```dart
> // Does not work regardless of whether this line is commented out
>   // FirebaseFunctions.instance.useFunctionsEmulator('localhost', 5001);
> ```

Usually, if you are running on firebase emulators, each flutterfire plugin has dedicate emulator to run on. Example: for fireabase_auth, we would use:

`await FirebaseAuth.instance.useAuthEmulator('localhost', 9099);`

And I wonder if that's the reason you are getting `An internal error has occurred, print and inspect the error details for more information.`

**Comment by hspinks** [comment]
@darshankawar no problem, I've tried running it with and without the emulator, and I'm still getting the same error. Here's my full `main_development.dart` (which is my entry point for the development flavor of the app):

```dart
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:cloud_functions/cloud_functions.dart';
import 'package:my_app/app.dart';
import 'package:my_app/config/firebase_options_dev.dart';
import 'package:firebase_auth/firebase_auth.dart';
import 'package:firebase_core/firebase_core.dart';
import 'package:flutter/material.dart';
import 'package:logger/logger.dart';

void main() async {
  // Recommended by FlutterFire
  WidgetsFlutterBinding.ensureInitialized();

  // Initializes Firebase from dev Firebase options
  await Firebase.initializeApp(
    options: DefaultFirebaseOptions.currentPlatform,
  );

  // Use local emulators
  await FirebaseAuth.instance.useAuthEmulator('localhost', 9099);
  FirebaseFirestore.instance.useFirestoreEmulator('localhost', 8080);
  FirebaseFunctions.instance.useFunctionsEmulator('localhost', 5001);

  // Set logger level to debug
  Logger.level = Level.debug;

  // Run app
  runApp(const MyApp());
}
```

My `app.dart` file looks like:

```dart
import 'package:cloud_functions/cloud_functions.dart';
import 'package:flutter/material.dart';

class MyApp extends StatefulWidget {
  const MyApp({Key? key}) : super(key: key);

  @override
  State<MyApp> createState() => _MyAppState();
}

class _MyAppState extends State<MyApp> {
  List fruit = [];

  @override
  Widget build(BuildContext context) {
    return MaterialApp(
      title: 'My App',
      home: Scaffold(
        body: Center(
          child: ListView.builder(
            itemCount: fruit.length,
            itemBuilder: (context, index) {
              return ListTile(
                title: Text('${fruit[index]}'),
              );
            },
          ),
        ),
        floatingActionButton: Builder(
          builder: (context) => FloatingActionButton.extended(
            onPressed: () async {
              // See index.js in .github/workflows/scripts for the example function we
              // are using for this example
              HttpsCallable callable = FirebaseFunctions.instance.httpsCallable(
                'listFruit',
                options: HttpsCallableOptions(
                  timeout: const Duration(seconds: 5),
                ),
              );

              try {
                final result = await callable();
                setState(() {
                  fruit.clear();
                  result.data.forEach((f) {
                    fruit.add(f);
                  });
                });
              } catch (e) {
                print(e);
                ScaffoldMessenger.of(context).showSnackBar(
                  SnackBar(
                    content: Text('ERROR: $e'),
                  ),
                );
              }
            },
            label: const Text('Call Function'),
            icon: const Icon(Icons.cloud),
            backgroundColor: Colors.deepOrange,
          ),
        ),
      ),
    );
  }
}
```

I've also tried minimalizing my `main_development.dart` to remove all of the other libraries (e.g. Firebase Auth, Firestore, and Logger) to isolate Cloud Functions, and I'm still seeing the same issue. It happens regardless of whether I use the emulator or not. I also have tried debugging the actual Cloud Function code by putting breakpoints and `console.log`s in, and they're never hit, so it seems like something is going wrong before it ever gets to the function. The other Firebase services I'm using in the app (Auth and Firestore) work great, both pointing at the local emulator or pointing at the Firebase project in the cloud, so it seems to be isolated to Cloud Functions.

**Comment by hspinks** [comment]
Update: this issue seems to be limited to iOS. Running the app for the web on Chrome works fine. I'll attempt to test on Android as well, but getting some build errors relating to my Android environment that I'll have to resolve first.

**Comment by darshankawar** [comment]
Thanks for further updates. I tried the code you shared [above](https://github.com/FirebaseExtended/flutterfire/issues/8037#issuecomment-1033935037) and  running it with and without firebase emulator, seems to be throwing below exception in the console which is same as you reported:

```
flutter: [firebase_functions[/17999]()] An internal error has occurred, print and inspect the error details for more information.
#0      StandardMethodCodec.decodeEnvelope
package:flutter/…/services/message_codecs.dart:607
#1      MethodChannel._invokeMethod
package:flutter/…/services/platform_channel.dart:167
<asynchronous suspension>
#2      MethodChannelHttpsCallable.call
package:cloud_functions_platform_interface/…/method_channel/method_channel_https_callable.dart:23
<asynchronous suspension>
#3      HttpsCallable.call
package:cloud_functions/src/https_callable.dart:35
<asynchronous suspension>
#4      _MyAppState.build.<anonymous closure>.<anonymous closure>
package:my_app/app.dart:42
<asynchronous suspension>
```

`cloud_functions: ^3.2.7`
`Channel stable, 2.10.1`

**Comment by russellwheatley** [comment]
Hey @hspinks, I've just tried on my iOS simulator and it works.

Are you sure your `<app-name>` is correct here:
```bash
✔  functions[us-central1-listFruit]: http function initialized (http://localhost:5001/<app-name>/us-central1/listFruit).
```  
i.e matches the [firebase projectId](https://github.com/FirebaseExtended/flutterfire/blob/master/packages/cloud_functions/cloud_functions/example/lib/firebase_config.dart#L25) in your app configuration?

**Comment by russellwheatley** [comment]
@hspinks Sorry, I cannot recreate it. If you want to create a repo with minimal amount of code for an app that I can spin up to reproduce the problem, I'll be happy to debug.

**Comment by hspinks** [comment]
@russellwheatley I ended up using a temporary workaround so I can keep my progress going forward: I'm writing documents to Firestore and using the `onCreate` trigger to run my cloud functions. I still intend to set up a repo with minimal code to repro, and will follow up here as soon as I do

**Comment by hspinks** [comment]
Update: this issue has somehow resolved itself without any changes to the code. I'd continued developing without focusing on this, and then when I returned to this original code today, it worked with no issues. I'll list out some of the things I'd done in the course of development of other features, but I didn't make any attempt to isolate this issue, so I have no clue which (if any) of these did the trick:

* Ran `flutter clean`
* Cleared my CocoaPods cache (e.g. `rm -rf ~/Library/Caches/CocoaPods`)
* Repaired a broken CocoaPods specs repository (e.g. `rm -rf ~/.cocoapods/repos/trunk`)
* Removed the line from my Podfile recommended [here](https://firebase.flutter.dev/docs/firestore/overview#4-optional-improve-ios--macos-build-times) for improving iOS/macOS build times with Firestore; then, removed my Podfile.lock, rebuilt pods, then re-added that pod

My best guess is that this had something to do with the CocoaPods cache and/or broken specs repository, but... 🤷🏻‍♂️

Hopefully some of those steps can help if someone else runs into this issue, sorry that I didn't do more to isolate the issue while it was occurring!

**Comment by russellwheatley** [comment]
Hey @hspinks, thanks for providing detailed feedback! I've had my fair share of iOS dependency problems so I feel your pain. Glad it is resolved!

**Comment by hspinks** [comment]
Update: I put the project down for a week, came back to it, with zero change whatsoever to my code base or environment, and this issue was occurring again. The only thing that changed during that week was that I'd quit out of VS Code, Xcode, and my iOS Simulator, then re-opened them. I tried each of my above steps, and no dice.

@russellwheatley is there anything else that you suggest for me to do to try to isolate the issue or help you replicate it? Clearly this is not an issue in the code, but since @darshankawar was able to replicate it, there is something going wrong somewhere. I'm really not sure how to proceed...

**Comment by hspinks** [comment]
Ok, set a bunch of breakpoints in the methods in the stack trace, will share here in case there is anything that you see that looks off...

<details>
<summary>package:cloud_functions_platform_interface-5.1.1/src/method_channel/method_channel_https_callable.dart:23</summary>

Set breakpoint at the line where the error occurs:
```dart
Object? result = await MethodChannelFirebaseFunctions.channel
    .invokeMethod('FirebaseFunctions#call', <String, dynamic>{
  'appName': functions.app!.name,
  'functionName': name,
  'origin': origin,
  'region': functions.region,
  'timeout': options.timeout.inMilliseconds,
  'parameters': parameters,
});
```

The data that is passed to the function:
<img width="775" alt="Screen Shot 2022-03-12 at 1 13 09 PM" src="https://user-images.githubusercontent.com/1484366/158035703-5d6f443c-b6a3-48d8-b4b2-74303edc02c4.png">
<img width="821" alt="Screen Shot 2022-03-12 at 1 13 17 PM" src="https://user-images.githubusercontent.com/1484366/158035692-41aea39f-befe-4a37-91b8-60cbf6e66461.png">

And here's the `MethodChannelFirebaseFunctions.channel`:
<img width="905" alt="Screen Shot 2022-03-12 at 1 08 56 PM" src="https://user-images.githubusercontent.com/1484366/158035741-049e60ff-a60d-453b-aa9d-4629a333ef09.png">
</details>

<details>
<summary>packages:flutter/src/~services/message_codecs.dart:607</summary>

Based on the code I see here, it looks like when reading the buffer, the first byte is not 0, which triggers reading error details.

Here's a screenshot that shows all of the details, including the error message, the error code, and the data array in the buffer (you can see that the first byte in the buffer is not 0):
<img width="1306" alt="Screen Shot 2022-03-12 at 1 20 52 PM" src="https://user-images.githubusercontent.com/1484366/158035864-d4499702-fec9-40ea-a4ec-fc0bffba6444.png">
</details>

It seems like something is going wonky when communicating over the bridge between Flutter and the iOS platform code, specifically when communicating with the `plugins.flutter.io/firebase_functions` plugin. What's interesting in watching these breakpoints is that they are hit multiple times (e.g. initializing Firebase core, initializing the app, etc) and seem to work totally fine, so the issue is isolated to cloud functions.

The only other thing that I can see in this is that it looks like the app name that is being passed down to these functions where the error is occurring is `"[DEFAULT]"` rather than the specific app ID. However, inspecting the rest of the FirebaseOptions, it looks like the app ID and all other app-specific info is there.

Please let me know if there is anything else that I can capture and share that would help track down the issue!

**Comment by hspinks** [comment]
Ok, sorry for the barrage of comments, but one more... all of a sudden, it's working again!

A little additional context - I use cloud functions in two places in the app: 
1. In my development environment/flavor, when configured to use emulators - in my `main_development.dart`, after initializing the app and pointing all my Firebase packages to the emulators, I hit a function to seed Firestore with data to help with development
2. I have a call in response to a user action

How it started working again:
1. I commented out the first call to seed Firestore so that it would not error at that point
3. I ran the app, which launched successfully
4. I triggered the second call in the iOS Simulator by the user action
5. I uncommented out the first call to seed Firestore
6. I live-restarted the app, which ran the seed call successfully

This would seem to indicate that there is something special/different about trying to call a cloud function while booting up the app, or soon after configuring Firebase (potential race condition?). However, in my prior attempts to debug, when I had implemented the example code, the function was being triggered by a user action, which happened long after app boot / Firebase configuration. So now, I am even more confused...

I really do apologize for the jumbled information in these comments - I've had a hard time finding the patterns of when/why the issue occurs, which is making it hard for me to communicate clearly.

**Comment by maganap** [comment]
OMG... I had already stopped, rebuilt, reverted to `cloud_functions: ^3.2.9`, cleared as much as I know, reinstalled pods, I even restarted my laptop... 

Although the error was 400, not 401, I thought about clearing auth just to test if it was related to user authentication...
I'm using `firebase_auth: ^3.3.6` as well. 

After it had just failed, in the same debug session **I signed out, made a hot restart, signed in, and it started working again.**

I'm laughing not to cry... 😅 

@hspinks can you reproduce?

**Comment by hspinks** [comment]
@maganap sadly, I have not figured out how to reliably reproduce... I'm in the same boat as you where the issue comes and goes even within the same debug session...

Including @darshankawar, that makes at least three of us that have run into this. There is nothing particularly special or custom about my setup, everything I'm using is pretty out-of-the-box. I'm also on an M1 Mac, using `firebase_auth: ^3.3.6` and `cloud_functions: ^3.2.9`. I'm on Flutter `2.10.1`, and I bootstrapped my app using [VeryGood CLI](https://github.com/VeryGoodOpenSource/very_good_cli) (though I also have seen this issue in an app where I used the standard `flutter create`). The other FlutterFire packages I'm using are Firestore, Dynamic Links, and FlutterFire UI (full dependency list below). I installed CocoaPods using Homebrew so [I wouldn't have to add `sudo arch -x86_64 gem install`](https://developer.apple.com/forums/thread/652822?answerId=697009022#697009022) for all of my CocoaPods stuff.

It seems to me like the issue occurs when calling platform-specific code from Flutter, so possibly something with how FlutterFire registers plugins? I'm not sure exactly how best to debug that, though, so any tips or pointers would be appreciated.

<details>
<summary>Dependencies</summary>

```yaml
dependencies:
  cloud_firestore: ^3.1.7
  cloud_functions: ^3.2.9
  firebase_auth: ^3.3.6
  firebase_core: ^1.12.0
  firebase_dynamic_links: ^4.0.6
  flutter:
    sdk: flutter
  flutterfire_ui: ^0.3.2
  intl: ^0.17.0
  logger: ^1.1.0
  share_plus: ^3.0.4
```
</details>
